doc/man/rlx.1
.TH "RLX" "1" "January 2016" "rlx 0.2.0" "User Commands"
.SH "NAME"
rlx -- Command line interface for couchdb.
.SH "USAGE"

.SP
rlx <command> [\-c] [\-\-color|\-\-no\-color] [\-c|\-\-compress] [\-\-log\-http]
.br
    [\-\-interactive|\-\-no\-interactive] [\-\-recursive] [\-\-noop] [\-\-force]
.br
    [\-\-help] [\-\-version] [\-s|\-\-server=<url>] [\-d|\-\-database=<name>]
.br
    [\-i|\-\-id=<id>] [\-a|\-\-attachment=<file>] [\-r|\-\-rev=<rev>]
.br
    [\-u|\-\-username=<name>] [\-p|\-\-password=<pass>] [\-o|\-\-output=<file>]
.br
    [\-f|\-\-file=<file>] [\-j|\-\-json=<json>] [\-t|\-\-template=<name>]
.br
    [\-q|\-\-query=<params...>] [\-h|\-\-header=<key:val...>]
.br
    [\-g|\-\-glob=<pattern...>] [\-\-log\-level=<level>] [\-\-destination=<url>]
.br
    [\-\-ddoc=<name>] [\-n|\-\-name=<name>] [\-\-feed=<type>] <args>
.SH "DESCRIPTION"
.PP
Designed for parity with the couchdb HTTP API, run \fBhelp <cmd>\fR for more information on individual commands, start an interactive session with \fBrlx i\fR.
.SH "COMMANDS"
.BL
.IP "\[ci]" 4
\fBadmin\fR: Manage server administrators.
.IP "\[ci]" 4
\fBuser, usr\fR: Manage server users.
.IP "\[ci]" 4
\fBdatabase, db\fR: Manage databases.
.IP "\[ci]" 4
\fBsecurity, sec\fR: Get or set database security.
.IP "\[ci]" 4
\fBdocument, doc\fR: Manage documents.
.IP "\[ci]" 4
\fBbulk\fR: Bulk document tasks.
.IP "\[ci]" 4
\fBattach, att\fR: Manage document attachments.
.IP "\[ci]" 4
\fBreplicate, repl\fR: Replicate a database.
.IP "\[ci]" 4
\fBbatch\fR: Batch file execution.
.IP "\[ci]" 4
\fBapplication, app\fR: Manage design document applications.
.IP "\[ci]" 4
\fBlocal, lcl\fR: Manage local documents.
.IP "\[ci]" 4
\fBlogin, in\fR: Login to a server.
.IP "\[ci]" 4
\fBlogout, out\fR: Logout of current session.
.IP "\[ci]" 4
\fBsession, sess\fR: Session authentication.
.IP "\[ci]" 4
\fBconfig, conf\fR: Manage server configuration.
.IP "\[ci]" 4
\fBinfo\fR: Print server information.
.IP "\[ci]" 4
\fBstats\fR: Print server statistics.
.IP "\[ci]" 4
\fBtasks\fR: Print active tasks. 
.IP "\[ci]" 4
\fBlog\fR: Tail server log file. 
.IP "\[ci]" 4
\fBrestart\fR: Restart a server.
.IP "\[ci]" 4
\fBuuids\fR: Print list of uuids.
.IP "\[ci]" 4
\fBedit\fR: Edit a document.
.IP "\[ci]" 4
\fBtpl\fR: Manage template files.
.IP "\[ci]" 4
\fBlint\fR: Lint javascript and json.
.IP "\[ci]" 4
\fBlevel, lvl\fR: Get or set the server log level.
.IP "\[ci]" 4
\fBwhoami\fR: Get current user information.
.IP "\[ci]" 4
\fBrunconf, rc\fR: Runtime configuration.
.IP "\[ci]" 4
\fBalias, as\fR: Manage alias shortcuts.
.IP "\[ci]" 4
\fBsystem, sys\fR: System utilities.
.IP "\[ci]" 4
\fBjson, j\fR: Grep JSON and more.
.IP "\[ci]" 4
\fBhttp, tt\fR: Low\-level HTTP requests.
.IP "\[ci]" 4
\fBinteractive, i\fR: Interactive REPL console.
.IP "\[ci]" 4
\fBhelp\fR: Show help for commands.
.EL
.SH "OPTIONS"
.BL
.IP "\[ci]" 4
\fB\-h, \-\-header=[key:val...]\fR: Additional HTTP headers.
.IP "\[ci]" 4
\fB\-\-[no]\-color\fR: Enable or disable terminal colors.
.IP "\[ci]" 4
\fB\-d, \-\-database=[name]\fR: Database name.
.IP "\[ci]" 4
\fB\-i, \-\-id=[id]\fR: Document identifier.
.IP "\[ci]" 4
\fB\-a, \-\-attachment=[file]\fR: Attachment file path.
.IP "\[ci]" 4
\fB\-r, \-\-rev=[rev]\fR: Document revision.
.IP "\[ci]" 4
\fB\-u, \-\-username=[name]\fR: Authentication username.
.IP "\[ci]" 4
\fB\-p, \-\-password=[pass]\fR: Authentication password.
.IP "\[ci]" 4
\fB\-o, \-\-output=[file]\fR: Write response to output file.
.IP "\[ci]" 4
\fB\-f, \-\-file=[file]\fR: Read JSON input from file.
.IP "\[ci]" 4
\fB\-j, \-\-json=[json]\fR: JSON string literal (overrides \-\-file).
.IP "\[ci]" 4
\fB\-t, \-\-template=[name]\fR: Name of a template file.
.IP "\[ci]" 4
\fB\-c, \-\-compress\fR: Compact JSON output format (zero indent). 
.IP "\[ci]" 4
\fB\-q, \-\-query=[params...]\fR: Query string parameters.
.IP "\[ci]" 4
\fB\-s, \-\-server=[url]\fR: Database server.
.IP "\[ci]" 4
\fB\-g, \-\-glob=[pattern...]\fR: File match pattern(s).
.IP "\[ci]" 4
\fB\-\-log\-level=[level]\fR: Set the log level.
.IP "\[ci]" 4
\fB\-\-log\-http\fR: Enable HTTP logs.
.IP "\[ci]" 4
\fB\-\-[no]\-interactive\fR: Disable interactive prompts.
.IP "\[ci]" 4
\fB\-\-recursive\fR: Read directories recursively.
.IP "\[ci]" 4
\fB\-\-noop\fR: Print db network requests, do not send them.
.IP "\[ci]" 4
\fB\-\-destination=[url]\fR: Copy destination.
.IP "\[ci]" 4
\fB\-\-force\fR: Force file overwrite.
.IP "\[ci]" 4
\fB\-\-ddoc=[name]\fR: Design document id.
.IP "\[ci]" 4
\fB\-n, \-\-name=[name]\fR: Design document function name (view|update|show|list).
.IP "\[ci]" 4
\fB\-\-feed=[type]\fR: Poll feed type. 
.IP "\[ci]" 4
\fB\-\-help\fR: Display this help and exit.
.IP "\[ci]" 4
\fB\-\-version\fR: Print version and exit.
.EL
.SH "EXPANSION"
.PP
Expansion is performed on the \fB\-s | \-\-server\fR option for absolute URLs and aliases. You may specify the database, document and revision such that the URL:

  \-s http://localhost:5984/data%2Fstore/path/document?rev=0\-1
.PP
Is expanded to the equivalent of:

  \-s http://localhost:5984 \-d data/store \-i path/document \-r 0\-1
.PP
Note that when referencing databases in the URL form you should URL encode the database name whilst it is not necessary for the \fB\-d\fR option.
.PP
If \fB\-\-database\fR, \fB\-\-id\fR, \fB\-\-rev\fR or \fB\-\-username\fR are specified they override any values extracted during expansion.
.PP
See rlx\-alias(1) for more information on alias expansion.
.SH "LOG"
.PP
All log output is sent to \fBstderr\fR, response documents are printed to \fBstdout\fR. You may set the program log level using the \fB\-\-log\-level\fR option. The default log level is \fBinfo\fR.
.PP
Enable logging of HTTP requests and responses by also specifiying \fB\-\-log\-http\fR option, any level specified using the \fB\-\-log\-level\fR option applies to the HTTP logger. When the HTTP log level is \fBinfo\fR basic request and response information is logged (HTTP verb, request URL and response status code), when the log level is \fBdebug\fR then query string parameters and headers will also be logged.
.SH "HEADERS"
.PP
Similar to \fBcurl(1)\fR arbitrary HTTP headers are accepted using the \fB\-h | \-\-header\fR option (note the short version is lowercase), you may specify as many headers as you like although \fBContent\-Type\fR and \fBAccept\fR are ignored, a warning is printed if you attempt to use them.
.PP
If the value of the \fB\-\-header\fR option is incorrect (bad syntax) an error is reported.
.PP
Use the HTTP logging option to inspect the request headers:

.SP
  rlx doc get \-s {server} \-d {database} \-i {document} \e
.br
    \-h 'if\-none\-match: "{revision}"' \-\-log\-http \-\-log\-level debug
.PP
The \fB\-\-rev\fR would be more concise in the above example, however it illustrates the ability to set and inspect headers.
.SH "QUERY\-STRING"
.PP
The \fBcouchdb\fR database server accepts many different query string parameters for different API calls, as such the query string parameter handling is flexible.
.PP
To specify query string parameters use the \fB\-q | \-\-query\fR option, this option is repeatable so you may specify each parameter as an individual option or combine the entire query string.
.PP
Some options map to query string parameters (for example \fB\-r | \-\-rev\fR) if you specify an option that maps to a query string parameter and the same parameter using \fB\-\-query\fR then the specific option value has precedence.
.PP
You may specify a leading \fB?\fR but it is unnecessary and not recommended.
.PP
To elucidate you can fetch document revision information with the \fBrevsinfo\fR shortcut command:

  rlx doc revsinfo \-s {server} \-d {database} \-i {id}
.PP
But you could also use \fB\-q | \-\-query\fR:

  rlx doc get \-s {server} \-d {database} \-i {id} \-q 'revsinfo=true'
.PP
An example of precedence:

  rlx doc get \-s {server} \-d {database} \-i {id} \-r {rev} \-q 'rev={revision}'
.PP
The value of \fB{rev}\fR will be used \fInot\fR \fB{revision}\fR.
.SH "HIGHLIGHT"
.PP
The program will attempt to syntax highlight JSON and javascript documents using either \fBsource\-highlight\fR or \fBpygmentize\fR. Document highlighting will not occur under the following conditions:
.BL
.IP "\[ci]" 4
Neither \fBsource\-highlight\fR or \fBpygmentize\fR is in \fB$PATH\fR.
.IP "\[ci]" 4
The \fBstdout\fR stream is not a \fBtty\fR (redirection).
.IP "\[ci]" 4
The \fB\-o | \-\-output\fR option is specified (output is a file).
.IP "\[ci]" 4
The \fB\-\-no\-color\fR option is specified (disables all ANSI escape sequences).
.IP "\[ci]" 4
The rc file \fBhighlight\fR section is invalid (does not contain json and js objects).
.IP "\[ci]" 4
The output to print is neither JSON or javascript (\-\-raw specified).
.EL
.SH "ENVIRONMENT"
.PP
The \fB$HOME\fR environment variable must be set in order for user configuration data to be loaded, typically for most developer environments this is not an issue however when using \fBrlx(1)\fR as part of infrastructure deployment you may need to ensure that \fB$HOME\fR is set.
.SH "FILES"
.PP
A debug log file (\fBrlx\-debug.log\fR) is written to the current working directory when an error occurs. The debug log file is written synchronously when the program exits.
.PP
The format of this file and under which circumstances it is written is controlled by the rc configuration section:

.SP
  "log": {
.br
    "debug": {
.br
      "clean": true,
.br
      "always": false,
.br
      "interactive": false,
.br
      "json": false,
.br
      "lines": false,
.br
      "indent": 2
.br
    }
.br
  }
.PP
The rules for when a debug log file are written are:
.BL

  1.  If \fBalways\fR write the file regardless of exit code or interactivity.

  2.  Otherwise write the file if an error occurs and it is not an interactive session.

  3.  If \fBinteractive\fR also write the file for interactive sessions with errors.
.EL
.PP
The log file by default is a plain text format designed to be quick to scan you can set \fBjson\fR to \fBtrue\fR for a dump of the log records array, if you also set \fBlines\fR to \fBtrue\fR then the output is a newline delimited list of stringified log records compatible with bunyan(1). The \fBindent\fR property only applies to JSON output when \fBlines\fR is not set.
.PP
The \fBclean\fR property indicates that the log file should be removed upon successful execution (zero exit code).
.PP
Input files are read with the following precedence: \fBstdin\fR, \fB\-\-json\fR and \fB\-\-file\fR. 
.PP
To read from \fBstdin\fR the special option (\-) must be specified, if data is written to \fBstdin\fR it must be a JSON document or an error is reported. When the \fB\-\-json\fR option is specified it must be a valid JSON string literal, be sure to enclose in quotes to prevent shell errors.
.PP
If both the \fB\-\-json\fR and \fB\-\-file\fR options are specified the JSON literal is preferred.
.PP
Files specified using the \fB\-\-file\fR option may be a file system path or URL. Relative paths are resolved according to the current working directory. For example:

.SP
  rlx lint \-f package.json
.br
  rlx lint \-f http://registry.npmjs.org
.br
  rlx lint \-j '{}'
.br
  cat package.json | rlx \- lint
.SH "HISTORY"
.PP
This program was originally implemented in bash shell script.
.PP
Bash was chosen for ease of readline integration and the ability to concisely pipe between programs amongst other features. However implementing a complex program in shell script is non\-trivial and it needed to rely on external languages for JSON support.
.PP
The original implementation whilst almost feature complete was deprecated in favour of a pure javascript version. In addition the original implementation started from a pure interactive REPL perspective with a view to implementing non\-interactive support later \- it never happened.
.PP
For the javascript program an inverse approach is taken, the REPL is the last feature to be implemented.
.SH "BUGS"
.PP
Report bugs to https://github.com/tmpfs/rlx/issues.
.SH "SEE ALSO"
.PP
rlx\-admin(1), rlx\-alias(1), rlx\-application(1), rlx\-attach(1), rlx\-batch(1), rlx\-bulk(1), rlx\-config(1), rlx\-database(1), rlx\-document(1), rlx\-edit(1), rlx\-http(1), rlx\-info(1), rlx\-interactive(1), rlx\-json(1), rlx\-level(1), rlx\-lint(1), rlx\-local(1), rlx\-log(1), rlx\-login(1), rlx\-logout(1), rlx\-runconf(1), rlx\-replicate(1), rlx\-restart(1), rlx\-security(1), rlx\-session(1), rlx\-stats(1), rlx\-system(1), rlx\-tasks(1), rlx\-tpl(1), rlx\-user(1), rlx\-uuids(1), rlx\-whoami(1)


lib/rlx.js
//process.env.CLI_TOOLKIT_DEBUG=true;
//process.env.CLI_TOOLKIT_MIDDLEWARE_REQUEST=true;
//process.env.CLI_INPUT_NO_READLINE_PATCH=true;

//var start = process.hrtime();

var path = require('path');

var util = require('util');
var querystring = require('querystring');
var circular = require('circular');

var cdb = require('cdb');

var glue = require('cli-interface');

//var diff = process.hrtime(start);
//console.log('requires took %d microseconds',
  //(diff[0] * 1e9 + diff[1]) / 1000);

var cli = require('cli-command'), help = cli.help;

var rc = require('cli-rc');
var logger = require('cli-logger');
var CommandInterface = glue.CommandInterface;
var ConverterMap = require('cli-converter-util').ConverterMap;

var boot = require('./boot');
var ready = require('./ready');
var before = require('./before');
var stringify = require('./util/stringify');
var resolve = require('./util/resolve');
var dirs = require('./request/dirs');
var expand = require('cli-property').expand;
var debuglog = require('./util/debug-log');
var merge = require('cli-util').merge;

var base = path.normalize(path.join(__dirname, '..'));

//process.stdin.resume();

var Relax = function() {
  CommandInterface.apply(this, arguments);
}

util.inherits(Relax, CommandInterface);

Relax.prototype.configure = function() {
  var file = path.join(__dirname, this.name() + '.md');
  var options = {
    commands: {
      log: {
        options: {
          offset: cli.types.integer,
          bytes: cli.types.integer
        }
      },
      uuids: {
        options: {
          count: cli.types.integer
        }
      }
    },
    options: {
      json: cli.types.json,
      //console.dir(cli.types.json);
      //file: cli.types.resource('-f', true, ['json', 'js']),
      file: new ConverterMap({
        default: cli.types.resource('-f', true, ['json', 'js']),
        attach: cli.types.resource('-f', true),
        application: cli.types.resource('-f', true)
      }),
      attachment: cli.types.string,
      feed: cli.types.enum(Object.keys(cdb.feeds)),
      logLevel: cli.types.enum(logger.keys)
    }
  }
  var conf = {
    start: {
      time: new Date(),
      cwd: process.cwd()
    },
    stdin: true,
    trace: process.env.NODE_ENV === 'devel',
    //trace: true,
    compiler: {
      input: [file],
      output: path.join(__dirname, this.name() + 'c.js'),
      definition: options,
      cache: process.env.NODE_ENV !== 'devel'
    },
    command: {
      dir: path.join(__dirname, 'command'),
      required: true,
      delimiter: /[\/:]/,
      before: before
    },
    error: {
      locales: path.join(__dirname, 'error', 'locales'),
      log: {
        print: false
      },
      intercept: function(req, next, e, err, parameters, source) {
        // TODO: we always need a list of errors for the request.
        //req.errors = req.errors || [];
        //req.errors.push(
          //{e: e, err: err, parameters: parameters, source: source });
        //

        //console.log('intercept called');

        if(!req.rc) return true;
        if(!req.rc.error.json) return true;
        // emit raise so we still capture the error
        this.emit('raise', e, this.errors, err, parameters, source);
        req.error.call(this, e, req, next);
        return false;
      }
    },
    manual: {
      dir: path.join(base, 'doc', 'man'),
      dynamic: process.env.NODE_ENV === 'devel'
    },
    env: {
      merge: false,
      cache: true,
      //cache:  process.env.NODE_ENV !== 'test',
      rcmerge: function(env, rc) {
        var expanded = expand(env);
        var merged = merge(expanded, rc);
        return env;
      },
      expand: {
        delimiter: '.'
      },
      native: {delimiter: ',', json: false}
    },
    rc: {
      merge: false,
      type: rc.JSON,
      cache: process.env.NODE_ENV !== 'test',
      //enabled: process.env.NODE_ENV !== 'test',
      base: base,
      name: '.' + this.name() + 'rc',
      resolve: resolve,
      keys: {
        append: 'searchPath'
      },
      home: function() {
        return dirs.home;
      },
      onload: function(req, next, rc, runcontrol) {
        var match
          , ptn = rc.alias.match
          , ignores = rc.history.ignores;

        try{
          match = new RegExp(ptn);
        }catch(e) {
          return req.error(this.wrap(
            this.errors.EREGEXP_COMPILE, [ptn, 'alias.match'], e), req, next);
        }

        rc.alias.match = match;

        if(!Array.isArray(ignores)) {
          ignores = rc.history.ignores = [ignores];
        }

        for(var i = 0;i < ignores.length;i++) {
          try {
            ignores[i] = new RegExp(ignores[i]);
          }catch(e) {
            return req.error(this.wrap(
              this.errors.EREGEXP_COMPILE,
              [ignores[i], 'history.ignores[' + i + ']'], e), req, next);
          }
        }

        //console.log('after rc load: %s', this.server);
        //console.dir(rc);

        next();
      }
    },
    variables: {
      prefix: '@',
      coerce: true,
      delimiter: ','
    },
    help: {
      width: 26
    },
    debug: {
      file: this.name() + '-debug.log'
    },
    boot: boot,
    ready: ready,
  };
  this
    .configure(conf)
    .usage();
}

Relax.prototype.use = function() {
  var opts = {level: logger.INFO, json: false};

  this
    .use(require('cli-mid-color'))
    .use(require('cli-mid-logger'), opts)
}

Relax.prototype.on = function() {

  var scope = this
    , conf = this.configure();

  if(process.env.NODE_ENV  === 'test') {
    process.setMaxListeners(4096);
  }

  process.on('exit', function onexit(code) {
    // write debug log file
    debuglog.call(scope, code, scope.configure().rc);
  })

  this
    .once('load', function(req) {
      //console.log('load %s', this.options().server.value());
      //this.reset();
      this
        .use(require('cli-mid-manual'))
        .help('--help')
        .version(null, null, 'Print version and exit');
    })
    .on('notify:database', function(req, next, arg, value) {
      if(this.database) {
        this.database = querystring.unescape(this.database);
      }
      if(conf.interactive && value === undefined) {
        delete this.database;
      }else{
        var validator = require('./startup/validator/db');
        return validator.call(this, req, next);
      }
      next();
    })
    .on('raise', function(err, errors) {
      var conf = this.configure();
      var errs = conf.errors = conf.errors || [];
      var o = {
        name: err.name,
        message: err.message,
        code: err.code,
        key: err.key,
        parameters: err.parameters || [],
        stack: err.toStackArray(),
        time: new Date()
      }
      errs.push(o);
    })
    .on('help:trailers', function ontrailers(doc, data, stream) {
      var i, cmd, col, conf = this.configure();
      var cmds = [
        {
          name: util.format('%s --help <cmd>', data.name),
          message: 'quick help on <cmd>'
        },
        {
          name: util.format('%s help <cmd>', data.name),
          message: 'documentation for <cmd>'
        },
        {
          name: util.format('%s i', data.name),
          message: 'launch interactive console'
        }
      ];

      if(conf.interactive) {
        cmds.pop();
      }

      if(doc.style === 'cmd') {
        doc.print(stream);
        for(i = 0;i < cmds.length;i++) {
          cmd = cmds[i];
          col = doc.toColumns.call(doc, cmd.name, cmd.message);
          doc.print(stream, col.left + col.right);
        }
      }

      if(doc.style !== 'man') {
        // overall footer
        doc.print(stream);
        doc.print(stream,
          util.format('%s@%s %s',
            data.name, data.version, path.dirname(__dirname)));
      }
    })
    .on('empty', function(help, version, req, next) {
      var conf = this.configure();
      help.call(this, 'cmd');
    })
    .once('complete', function(req) {


      var conf = this.configure();

      // error occured during bootup of interactive console
      if(conf.interactive && conf.errors.length) {
        process.exit(conf.errors.pop().code);
      }

      // standard end of execution
      if(process.env.NODE_ENV !== 'test' && !conf.interactive) {
        process.exit(0);
      }
    })
}

module.exports = function(pkg, name, description) {
  return new Relax(pkg, name, description);
}
